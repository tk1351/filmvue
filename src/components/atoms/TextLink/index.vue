<script setup lang="ts">
import { computed } from 'vue'

const props = withDefaults(
  defineProps<{
    title?: string
    href?: string
    color?: string
    disabled?: boolean
  }>(),
  {
    title: '',
    href: '',
    color: 'primary',
    disabled: false,
  }
)

const classObject = computed(() => {
  return {
    '-primary': props.color === 'primary',
    '-secondary': props.color === 'secondary',
    '-disabled': props.disabled,
  }
})

interface Emits {
  (eventName: 'click:link', event: Event): void
}
const emit = defineEmits<Emits>()

const handleClick = (event: Event) => {
  emit('click:link', event)
}
</script>

<template>
  <a :href="href" class="text-link" :class="classObject" @click="handleClick">{{
    title
  }}</a>
</template>

<style scoped lang="scss">
.text-link {
  &:hover {
    opacity: 0.8;
  }
}

.text-link.-primary {
  color: $main-color;
}

.text-link.-secondary {
  color: #ffffff;
}

.text-link.-disabled {
  pointer-events: none;
}
</style>
