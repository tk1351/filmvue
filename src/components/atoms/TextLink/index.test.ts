import { describe, expect, it } from 'vitest'
import { mount } from '@vue/test-utils'
import TextLink from './index.vue'

const wrapperFactory = () => {
  return mount(TextLink, {
    props: {
      title: 'title',
      href: 'href',
    },
  })
}

describe('Testing TextLink Component', () => {
  it('TextLinkコンポーネントが正しく表示されていること', () => {
    const wrapper = wrapperFactory()
    expect(wrapper.element).toMatchSnapshot()
  })

  describe('Props', () => {
    describe('title', () => {
      it('渡された値が正しく表示されていること', () => {
        const expectedTitle = 'expectedTitle'
        const wrapper = mount(TextLink, {
          props: { title: expectedTitle },
        })
        expect(wrapper.text()).toBe(expectedTitle)
      })
    })
    describe('href', () => {
      it('href属性に正しく値が渡っていること', () => {
        const expectedHref = 'expectedHref'
        const wrapper = mount(TextLink, {
          props: { href: expectedHref },
        })
        expect(wrapper.attributes('href')).toBe(expectedHref)
      })
    })
    describe('color', () => {
      it('colorがprimaryの場合、.-primaryクラスが付与されていること', () => {
        const wrapper = mount(TextLink, {
          props: { color: 'primary' },
        })
        expect(wrapper.classes('-primary')).toBe(true)
      })
      it('colorがsecondaryの場合、.-secondaryクラスが付与されていること', () => {
        const wrapper = mount(TextLink, {
          props: { color: 'secondary' },
        })
        expect(wrapper.classes('-secondary')).toBe(true)
      })
    })
    describe('disabled', () => {
      it('disabledがtrueの場合、.-disabledクラスが付与されていること', () => {
        const wrapper = mount(TextLink, {
          props: { disabled: true },
        })
        expect(wrapper.classes('-disabled')).toBe(true)
      })

      it('disabledがfalseの場合、.-disabledクラスが付与されていないこと', () => {
        const wrapper = mount(TextLink, {
          props: { disabled: false },
        })
        expect(wrapper.classes('-disabled')).toBe(false)
      })
    })
  })

  describe('Event', () => {
    it('リンクをクリックした時、click:linkイベントが一度発動すること', async () => {
      const wrapper = wrapperFactory()
      await wrapper.trigger('click')
      expect(wrapper.emitted('click:link')).toHaveLength(1)
    })
  })
})
