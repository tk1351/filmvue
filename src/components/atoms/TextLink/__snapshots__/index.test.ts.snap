// Vitest Snapshot v1

exports[`Testing TextLink Component > TextLinkコンポーネントが正しく表示されていること 1`] = `
<a
  class="text-link -primary"
  data-v-8475e6b2=""
  href="href"
>
  title
</a>
`;
